### PR TITLE
Error on images from docker.pkg.github.com

### LINKED ISSUE #113: Warn about containers pulling from docker.pkg.github.com
Upgraded my DOKS 1.19 to 1.20 cluster yesterday and found out the hard way `docker.pkg.github.com` can't be pulled by `containerd`.

### PR DESCRIPTION
**Pull request by WyriHaximus**
Resolves #113 

Since Kubernetes 1.20 containerd is used instead of Docker as the container runtime.
Containerd is due to protocol version differences/support/mismatch unabled to 
pull images hosted at docker.pkg.github.com. The new check in this commit will 
error when it finds an image from that registry, and suggests to use ghcr.io, 
which is also hosted and operated by GitHub, and the successor of 
docker.pkg.github.com.

Refs:
* https://github.com/containerd/containerd/issues/3291#issuecomment-683700425
* https://docs.github.com/en/packages/guides/migrating-to-github-container-registry-for-docker-images#domain-changes

### CODE REVIEW

File: checks/basic/docker_pkg_github_com_registry.go
Comment (timoreimann) [inline checks/basic/docker_pkg_github_com_registry.go:82]: Wouldn't it suffice to exact-match on the result from `reference.Domain(namedRef)`?

File: checks/basic/docker_pkg_github_com_registry.go
Comment (timoreimann) [inline checks/basic/docker_pkg_github_com_registry.go:42]: I wonder if we should put the check in the `doks` group where we know for sure that it's going to be an issue for users eventually (i.e., when they migrate to 1.20). Other providers/clusters may use CRI-O which I don't know whether it's affected as well. 🤔

File: checks/basic/docker_pkg_github_com_registry.go
Comment (timoreimann) [inline checks/basic/docker_pkg_github_com_registry.go:66]: What do you think about a method name like `checkDomainName` (since we do not really check the tags I believe)?

**Review by timoreimann — COMMENTED** [review]
Thanks for filing the issue and submitting a PR right away. 💙 

I left a few comments/questions. @varshavaradarajan / @adamwg would appreciate your review as well.

File: checks/basic/docker_pkg_github_com_registry.go
Comment (WyriHaximus) [reply checks/basic/docker_pkg_github_com_registry.go:42]: @timoreimann Is there a way to detect which version you're running the linter against? So we could make it a warning for everything below 1.20, and an error for 1.20 and up.

It's up to you where we should put it, if others are also using `containerd` on their kaas they will run into the same issue. If there is a way to detect if it runs against a DOKS cluster or not, we could do something with that and change the message.

Comment (timoreimann) [reply checks/basic/docker_pkg_github_com_registry.go:41]: >Is there a way to detect which version you're running the linter against? So we could make it a warning for everything below 1.20, and an error for 1.20 and up.

Not right now to my knowledge, although including that information shouldn't be too hard since Kubernetes provides a discovery endpoint with the server version. We could retrieve that information on startup and make it available to all checks.

The 1.20 version is just critical for DOKS: I think the official removal for Docker runtime support only happens in a few releases, but DOKS went ahead and made the switch sooner. So strictly speaking, we should only error out if we're on DOKS 1.20, which we neither provide as an input to clusterlint today but that could be determined heuristically based on certain labels we put on Node objects.

To complete the wish list, it'd be even greater if we could make the target Kubernetes version a configurable parameter to clusterlint: that way, we could not only error out on DOKS 1.20 but also on DOKS 1.19 right before a customer intends to upgrade to the version that's impacted. The latter is the very scenario you ran into a believe, so it'd be beneficial to get that ahead of time warning from clusterlint before things start to fall apart.

So that's a lot I'm asking for, but I don't want to block the PR on dependent fundamental changes needed to reap the biggest benefit -- the check is already very useful on its own. My gut feeling is that we could put it in the DOKS group for now (either as a warning or an error) where we know it's going to apply for sure eventually, and file support tickets to track the discussed work. Once those our complete, the check could be updated to make it apply in more specific contexts.

I'd like to wait for some of my colleagues who get to spend more time with clusterlint before making any decisions, but do let me know what you think.

Comment (WyriHaximus) [reply checks/basic/docker_pkg_github_com_registry.go:41]: > > Is there a way to detect which version you're running the linter against? So we could make it a warning for everything below 1.20, and an error for 1.20 and up.
> 
> Not right now to my knowledge, although including that information shouldn't be too hard since Kubernetes provides a discovery endpoint with the server version. We could retrieve that information on startup and make it available to all checks.
> 

That would be awesome :+1: . Does it also reveal if `docker` or `containerd` is used as the container runtime? Because if it is, we could make the check even more precise.

> The 1.20 version is just critical for DOKS: I think the official removal for Docker runtime support only happens in a few releases, but DOKS went ahead and made the switch sooner. So strictly speaking, we should only error out if we're on DOKS 1.20, which we neither provide as an input to clusterlint today but that could be determined heuristically based on certain labels we put on Node objects already.
> 

Well thank you for pushing forward in that regard :+1: 

> To complete the wish list, it'd be even greater if we could make the target Kubernetes version a configurable parameter to clusterlint: that way, we could not only error out on DOKS 1.20 but also on DOKS 1.19 right before a customer intends to upgrade to the version that's impacted. The latter is the very scenario you ran into a believe, so it'd be beneficial to get that ahead of time warning from clusterlint before things start to fall apart.
> 

Yes exactly! Warning on the upgrade to 1.19, and error on the upgrade to 1.20. Honestly? I would just not make it possible to upgrade to 1.20 when this check triggers.

> So that's a lot I'm asking for, but I don't want to block the PR on dependent fundamental changes needed to reap the biggest benefit -- the check is already very useful on its own. My gut feeling is that we could put it in the DOKS group for now (either as a warning or an error) where we know it's going to apply for sure eventually, and file support tickets to track the discussed work. Once those our complete, the check could be updated to make it apply in more specific contexts.
> 
> I'd like to wait for some of my colleagues who get to spend more time with clusterlint before making any decisions, but do let me know what you think.

I agree those things should be in the cluster linter, but also agree that with the current impact of DOKS moving ahead with the change we should merge thing in. Probably with a few more changes.

**Comment by WyriHaximus** [comment]
Meh moving it to `doks` didn't go as planned, will have another look later today:
Screenshot from 2021-02-08 13-38-46

**Comment by WyriHaximus** [comment]
Managed to work around it, but would rather somehow used the existing file instead of duplicating it.

**Comment by adamwg** [comment]
:wave: Thanks for taking this on @WyriHaximus!

Two high-level suggestions:

1. I'd suggest that we put this check in a new group, `containerd`, which would be applicable to any cluster using containerd as its runtime, and also include it in the `doks` group. The directory layout is a little arbitrary (and doesn't actually impact which group(s) checks end up in in the tool) so feel free to move/refactor as needed.
2. We already fetch the `Node`s from the cluster, so the new check could check whether any nodes in the cluster are using containerd (via the [`NodeSystemInfo.ContainerRuntime`](https://github.internal.digitalocean.com/digitalocean/cthulhu/pull/55887) field). This would be better than using the cluster/node version, since (as you and Timo discussed) non-DOKS 1.20 clusters coudl still be using docker.

I'll take a detailed look at the change as well and leave any suggestions there, but wanted to write down these high-level things first.

**Comment by WyriHaximus** [comment]
> 1. I'd suggest that we put this check in a new group, `containerd`, which would be applicable to any cluster using containerd as its runtime, and also include it in the `doks` group. The directory layout is a little arbitrary (and doesn't actually impact which group(s) checks end up in in the tool) so feel free to move/refactor as needed.

Moved it to the `containerd` group.

> 2. We already fetch the `Node`s from the cluster, so the new check could check whether any nodes in the cluster are using containerd (via the [`NodeSystemInfo.ContainerRuntime`](https://pkg.go.dev/k8s.io/api/core/v1#NodeSystemInfo) field). This would be better than using the cluster/node version, since (as you and Timo discussed) non-DOKS 1.20 clusters coudl still be using docker.
> 

Cool, but that is for current nodes rights? And not the nodes that will be spun up correct? So if you're migrating from `docker` to `containerd` this specific check wouldn't know you're about to do that as the currently `ContainerRuntime` it will get is `docker` and not `containerd`.

> I'll take a detailed look at the change as well and leave any suggestions there, but wanted to write down these high-level things first.

Looking forward to it :+1: .

**Comment by adamwg** [comment]
> > 1. We already fetch the `Node`s from the cluster, so the new check could check whether any nodes in the cluster are using containerd (via the [`NodeSystemInfo.ContainerRuntime`](https://pkg.go.dev/k8s.io/api/core/v1#NodeSystemInfo) field). This would be better than using the cluster/node version, since (as you and Timo discussed) non-DOKS 1.20 clusters coudl still be using docker.
> 
> Cool, but that is for current nodes rights? And not the nodes that will be spun up correct? So if you're migrating from `docker` to `containerd` this specific check wouldn't know you're about to do that as the currently `ContainerRuntime` it will get is `docker` and not `containerd`.

Ah, yes, good point :man_facepalming:

File: checks/containerd/docker_pkg_github_com_registry.go
Comment (adamwg) [inline checks/containerd/docker_pkg_github_com_registry.go:84]: It would be good to specifically call out that this is a containerd limitation. Looking at [the GH docs](https://docs.github.com/en/packages/guides/migrating-to-github-container-registry-for-docker-images) it looks like the user needs to re-push images to ghcr.io (not just replace the string in their image name). Given that, I'd suggest we simplify the message here and link to the GH doc from the check documentation.

```suggestion
				Message:  fmt.Sprintf("containerd can't pull images from docker.pkg.github.com, used by container '%s'", container.Name),
```

**Review by adamwg — COMMENTED** [review]
Overall this looks good to me. Please also add the new check to [checks.md](https://github.com/digitalocean/clusterlint/blob/master/checks.md), preferably with a link to the relevant containerd issue and GitHub migration doc.

File: checks/containerd/docker_pkg_github_com_registry.go
Comment (WyriHaximus) [reply checks/containerd/docker_pkg_github_com_registry.go:84]: Which is in `checks.md` right? Was already working on that after your pointed that file out.

File: checks/containerd/docker_pkg_github_com_registry.go
Comment (WyriHaximus) [reply checks/containerd/docker_pkg_github_com_registry.go:41]: Done :+1:

File: checks/containerd/docker_pkg_github_com_registry.go
Comment (adamwg) [reply checks/containerd/docker_pkg_github_com_registry.go:84]: Yep, exactly :+1:

Comment (WyriHaximus) [reply checks/containerd/docker_pkg_github_com_registry.go]: Cool, writing that now. Including links to the Github docs and the `containerd` issue

**Comment by WyriHaximus** [comment]
> Overall this looks good to me. Please also add the new check to [checks.md](https://github.com/digitalocean/clusterlint/blob/master/checks.md), preferably with a link to the relevant containerd issue and GitHub migration doc.

Done :+1:

File: checks.md
Comment (adamwg) [inline checks.md:785]: Nit: suggest "hostname" instead of "package" here.

File: checks.md
Comment (adamwg) [inline checks.md:794]: This should be `docker.pkg.github.com/...` to actually trigger the check, right?

Comment (WyriHaximus) [reply checks.md:794]: Ah yes good catch :+1:

File: checks.md
Comment (WyriHaximus) [reply checks.md:785]: Changed it to `package registry`. It isn't a simple hostname change. I've images on both, and they are different product within GitHub with different possibilities. `docker.pkg.github.com/OWNER/REPO/IMAGE:tag` vs `ghcr.io/OWNER/IMAGE:tag` is an externally visible difference.

Comment (adamwg) [reply checks.md]: Ah, yes, good point - the new wording is good.

**Review by adamwg — APPROVED** [review]
Thanks again for the contribution! This looks great.

**Review by adamwg — APPROVED** [review]
Thanks again for the contribution! This looks great.

**Comment by WyriHaximus** [comment]
One question tho, why isn't circleci running for this PR?

**Comment by adamwg** [comment]
That's what I was wondering :smile: @varshavaradarajan Any idea?

**Comment by varshavaradarajan** [comment]
The webhook payload isn't being delivered. I tried redelivering it as well. It's failing with 400. 
@WyriHaximus - Can you rebase and push again if you haven't?

**Comment by WyriHaximus** [comment]
@varshavaradarajan done :+1:

**Comment by adamwg** [comment]
Not sure what's going on with CircleCI here, but everything it runs is passing locally for me so I think we're fine to merge.

**Comment by WyriHaximus** [comment]
Might be a permissions issue as I don't have write access to this repo and the merge ran file.

**Comment by varshavaradarajan** [comment]
We've had external contributors before and circleCI has run fine. The webhook payload isn't getting delivered at all for this. :( We may need to remove and add the circle ci webhook if the problem persists. But looks like all the tests passed on master after the merge.

**Comment by MattIPv4** [comment]
Hey, @WyriHaximus - Thanks a ton for this sweet PR! 😄

Would you please shoot me an email when you get a chance?
mcowley at digitalocean dot com 🎉

**Comment by WyriHaximus** [comment]
@MattIPv4 Done 👍

**Pull request merged**
